### DISCUSSION TITLE
Crazyflie 2.1 Assist mode flips the flie over

### DISCUSSION THREAD

**Discussion by jn887** [bitcraze/discussions#1009] [Q&A] [answered]
Hi,  I have a Crazyflie (CF) 2.1 where normal flight with a PS4 controller flies it as expected, i.e. Thrust lifts the CF vertically perfectly.  When I press the controller button to active Assist Mode however, the CF immediately flies off sideways and flips itself over.  It always flips and flies off to the same side such that it seems like the M1 and M4 motors are lifting much more than the M2 and M3 motors, i.e. the CF flips over with the M1 and M4 going over the M2 and M3 side.  

I have another CF where this behavior does not happen, the manual vertical thrust and Assist Mode buttons on the same controller operate the other CF as expected, it goes up and stays in place.  

One thing I noticed is that the client screen for the faulty CF drop-down Assist mode menu is slightly different than the same screen for the working CF.  For the faulty one, the Assist mode "height hold" and "hover" options are 'grayed out' and I cannot select them.   While the working CF, the Assist mode drop down menu options are selectable except for "Altitude hold" which is grayed out.  

The flip-over behavior I described above is with "position hold" selected.  Altitude hold does not seem to do anything on faulty CF.   Screenshots of the two Assist Mode menus for the two CF's are attached. 

Faulty Assist, Height Hold and Hover are not selectable:
![CF3_faultyAssist](https://github.com/bitcraze/discussions/assets/148887385/e70b6858-d5f2-43ec-93df-57368cef11a0)

Working Assist, Altitude hold not selectable:
![CF2_workingAssist](https://github.com/bitcraze/discussions/assets/148887385/4f7279bc-63a7-4704-a5b6-fedac65efa5b)

I made sure to load the most recent firmware with the deck on both CFs.

Thanks for any guidance.

**Comment by jn887** [comment]
Hi Bitcraze Team,

FYI, I just tried swapping the PMW 3901 optical flow deck (Flow v2) between these two CFs, and it seems like the flipping behavior follows a specific Flow v2.  I.e. Both CFs will work fine (manually and in Assist) with one Flow v2, while the same CF with our second Flow v2 deck does the flipping with Assist behavior.  

In the client console, I'm also seeing this ESTKALMON error a lot more and it gets 'stuck' with that error and doesn't go into "ready to fly" mode with the seemingly faulty Flow v2 deck vs. the functional Flow v2.   The functional Flow v2 deck also gets the same ESTKALMON error, but it seems to happen less frequently, and it is able to recover to "ready to fly" while the faulty one stays with the ESTKALMON error.

I also ran the propeller test on this one CF and it keeps failing for M2 regardless of which Flow v2 deck I'm using.  I'm not sure if this is related or how to fix this either.

Is there a fix for this?  Thank you so much for any help.


CF with seemingly faulty Flow v2: 
![image](https://github.com/bitcraze/discussions/assets/148887385/7e7877d2-902e-4c07-8885-3565235dee2e)


Same CF with functional Flow v2: 
![image](https://github.com/bitcraze/discussions/assets/148887385/62176ec9-f1db-4531-bcb5-85c48ad5d153)

Propeller test results with functional Flow v2:
![image](https://github.com/bitcraze/discussions/assets/148887385/49704547-4ded-4c2e-a515-26cfb4a9e68c)

**Comment by knmcguire** [comment, marked as answer]
Alright let's dissect it a bit because there are multiple things going on and it is a bit confusing.

1. The first question was that you were seeing weird behavior with assist mode, not understanding why some parts are grayed out and why it is flipping over.
2. Then you noticed that one of the flow deck didn't work properly and that the console was indicating warnings about the estimator and that it couldn't fly (another one was working properly
3. Then mentioned that one of the propeller test of the motors didn't pass.

Here are some explanations:
1- The hover and height mode can only be selected with a flow deck (or z-ranger). Then probably around that time, the flowdeck was not attaching properly. I do find it weird that the position option was there but that requires a bit more debugging I think
2- So the warning of the kalman estimator rate being off is kind of normal. The estimator usually needs a bit of time to converge. The fact that you are getting it more with the faulty one is interesting though. Have you looked dat the [motion.deltax/deltay](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#motion) in [a plotter tab](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/plotter_tab/) when moving it over a texture rich area?
3- The motor tests does not have anything to do with the flowdeck itself, but it says something about the flight performance. The fact that M2 is failing, probably means 2 things: Your propeller is [not well balanced](https://www.bitcraze.io/documentation/tutorials/balancing-propellers/), maybe due to some crashes or two. Or, the shaft of the motor is bended. Eitherway, you can improve it by replacing those parts, and then your flight performance will improve.


PS. Just a couple of things when asking for help on this forum. Please use code blocks for text like the console like this one 
![image](https://github.com/bitcraze/discussions/assets/11061260/a1439546-9f1f-47d6-b921-fc320e73e328). Also, try to compartilize the questions. It can all be confusing when you are starting out and a lot of things are happening, but for us, it is better if you do 1 discussion per issue you are having. We rather have a lot of small ones than a few big ones :) Check out the discussion rules here: https://github.com/orgs/bitcraze/discussions/55

**Reply by knmcguire** [reply]
Also keep an eye on this issue as well: https://github.com/bitcraze/crazyflie-firmware/issues/1329. I've posted a couple of tips to maybe solve your issue with the flowdeck, but I have been noticing some instability issues with the flowdeck with latest releases.

**Reply by jn887** [reply]
Hi Kimberly,

Thank you so very much for all of your very helpful guidance!  And also for kindly letting me know the discussion rules, I apologize, I'll make sure to follow them going forward.

As for this/these issues with the flipping over and strange flow deck behavior, literally all of your tips helped me resolve them.

1 - I tested the two flow decks and the two CFs being extra mindful of the electrical connection.  Our header pins on that side are very short and some slid to the opposite side and was probably creating an intermittent/poor contact.  Making sure there was a good contact did resolve what I thought was a faulty deck.  The two flow decks are now behaving identically on both CFs.   Waving/shaking the CF also does seem to improve the flight control in terms of stability and I'll get longer header pins to help with moving it away from GND and also to help our attach/contact issue.

3 - I replaced all four motors on the CF that was flipping over itself and that seems to have resolved that issue.  That CF is flying very well now.

2 - I'll start looking at the motion.deltax/deltay info, but these two CFs are behaving very well now so I'm all set.

Thank you so much for your incredibly prompt and helpful support!  You guys are fantastic.

**Reply by knmcguire** [reply]
Great! Yes good motors are essential for good flight :) a good position estimate too but that was not your issue apparently.
